<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.22.0 by Michael Rose
  Copyright 2013-2020 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="ko" class="no-js">
  <head>
    <style> 
      ::-webkit-scrollbar{ 
        width: 10px;
        height: 10px;
      }

      ::-webkit-scrollbar-track {
        width: 0px;
        background-color: #626262;
        /* border-radius: 5px; */
      }

      ::-webkit-scrollbar-thumb {
        width: 0px;
        background-color: #E2E2E2;
        border-radius: 5px;
      }

      ::-webkit-scrollbar-thumb:hover {
        width: 10px;
        height: 20px;
        /* background-color: rgba(190, 190, 190, 0.2); */
        background-color: #A2A2A2;
        border-radius: 5px;
      }

      ::-webkit-scrollbar-track:hover {
        width: 10px;
        /* background-color: rgba(150, 150, 150, 0.1); */
        background-color: #626262;
        border-radius: 5px;
        /* background: transparent; */
        /* border-radius: 10px; */
      }

      ::-webkit-scrollbar-button:start:decrement,::-webkit-scrollbar-button:end:increment {
          width:0px;
          height: 0px;
          /* background-color: rgb(14, 221, 24); */
          /* border-radius: 50%; */
      }
    </style>
    
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>[Rejected] PEP 490 - Chain exceptions at C level | secrett2633</title>
<meta name="description" content="Python Enhancement Proposal 490: ‘Chain exceptions at C level’에 대한 한국어 번역입니다.">


  <meta name="author" content="secrett2633">
  
  <meta property="article:author" content="secrett2633">
  


<meta property="og:type" content="article">
<meta property="og:locale" content="ko_KR">
<meta property="og:site_name" content="secrett2633's blog">
<meta property="og:title" content="[Rejected] PEP 490 - Chain exceptions at C level">
<meta property="og:url" content="http://localhost:4000/python/pep/490/">


  <meta property="og:description" content="Python Enhancement Proposal 490: ‘Chain exceptions at C level’에 대한 한국어 번역입니다.">







  <meta property="article:published_time" content="2025-09-26T22:35:41+09:00">



  <meta property="article:modified_time" content="2025-09-26T22:35:41+09:00">



  

  


<link rel="canonical" href="http://localhost:4000/python/pep/490/">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": "secrett2633",
      "url": "http://localhost:4000/"
    
  }
</script>






<!-- end _includes/seo.html -->



  <link href="/feed.xml" type="application/atom+xml" rel="alternate" title="secrett2633's blog Feed">


<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->


    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->


    <link rel="icon" type="image/png" sizes="32x32" href="https://secrett2633.github.io/assets/images/favicon/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="https://secrett2633.github.io/assets/images/favicon/favicon-16x16.png">
    <meta name="msapplication-TileColor" content="#ffc40d">
    <meta name="theme-color" content="#ffffff">
  </head>

  <body class="layout--single">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          secrett2633's blog
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="https://github.com/secrett2633">GitHub</a>
            </li></ul>
        
        <button class="search__toggle" type="button">
          <span class="visually-hidden">Toggle search</span>
          <svg class="icon" width="16" height="16" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 15.99 16">
            <path d="M15.5,13.12L13.19,10.8a1.69,1.69,0,0,0-1.28-.55l-0.06-.06A6.5,6.5,0,0,0,5.77,0,6.5,6.5,0,0,0,2.46,11.59a6.47,6.47,0,0,0,7.74.26l0.05,0.05a1.65,1.65,0,0,0,.5,1.24l2.38,2.38A1.68,1.68,0,0,0,15.5,13.12ZM6.4,2A4.41,4.41,0,1,1,2,6.4,4.43,4.43,0,0,1,6.4,2Z" transform="translate(-.01)"></path>
          </svg>
        </button>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="https://schema.org/Person">

  

  <!-- 2022.02.17 author content hidden -->
  <!-- <div class="author__content">
    
      <h3 class="author__name" itemprop="name">secrett2633</h3>
    
    
  </div> -->

  <div class="author__urls-wrapper">
    <!-- <button class="btn btn--inverse">Follow</button> -->
    <ul class="author__urls social-icons">
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs" rel="nofollow noopener noreferrer">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
    
      
      
      
      
    
    
      

<nav class="nav__list">
  
  <input id="ac-toc" name="accordion-toc" type="checkbox" />
  <label for="ac-toc">Toggle Menu</label>
  <ul class="nav__items">
    
      <li>
        
          <!-- title -->
              <span class="nav__sub-title">Backend</span>
              <hr>
        

        
        <ul>
          
            <li><a href="/backend/django/">Django</a></li>
          
            <li><a href="/backend/logging/">Logging</a></li>
          
        </ul>
        
      </li>
    
      <li>
        
          <!-- title -->
              <span class="nav__sub-title">Python</span>
              <hr>
        

        
        <ul>
          
            <li><a href="/python/pep/">PEP</a></li>
          
        </ul>
        
      </li>
    
      <li>
        
          <!-- title -->
              <span class="nav__sub-title">AI/ML</span>
              <hr>
        

        
        <ul>
          
            <li><a href="/ai/llm/">LLM</a></li>
          
            <li><a href="/ai/review/">Review</a></li>
          
        </ul>
        
      </li>
    
      <li>
        
          <!-- title -->
              <span class="nav__sub-title">DevOps</span>
              <hr>
        

        
        <ul>
          
            <li><a href="/devops/nginx/">Nginx</a></li>
          
            <li><a href="/devops/docker/">Docker</a></li>
          
            <li><a href="/devops/safeline/">SafeLine</a></li>
          
            <li><a href="/devops/jenkins/">Jenkins</a></li>
          
            <li><a href="/devops/github-actions/">GitHub Actions</a></li>
          
            <li><a href="/devops/aws/">AWS</a></li>
          
        </ul>
        
      </li>
    
      <li>
        
          <!-- title -->
              <span class="nav__sub-title">etc</span>
              <hr>
        

        
        <ul>
          
            <li><a href="/etc/me/">Me</a></li>
          
            <li><a href="/etc/chrome-extension/">Chrome Extension</a></li>
          
        </ul>
        
      </li>
    
  </ul>
</nav>

    
  
  </div>



  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="[Rejected] PEP 490 - Chain exceptions at C level">
    <meta itemprop="description" content="Python Enhancement Proposal 490: ‘Chain exceptions at C level’에 대한 한국어 번역입니다.">
    <meta itemprop="datePublished" content="2025-09-26T22:35:41+09:00">
    <meta itemprop="dateModified" content="2025-09-26T22:35:41+09:00">

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">[Rejected] PEP 490 - Chain exceptions at C level
</h1>
          

  <p class="page__meta">
    
      
      <span class="page__meta-date">
        <i class="far fa-calendar-alt" aria-hidden="true"></i>
        <time datetime="2025-09-26T22:35:41+09:00">September 26, 2025</time>
      </span>
    

    

    
  </p>


        </header>
      

      <section class="page__content" itemprop="text">
        
          <aside class="sidebar__right sticky">
            <nav class="toc">
              <header><h4 class="nav__title"><i class="fas fa-file-alt"></i> On this page</h4></header>
              <ul class="toc__menu"><li><a href="#pep-490--c-수준에서-예외-연결-chain-exceptions-at-c-level">PEP 490 – C 수준에서 예외 연결 (Chain exceptions at C level)</a><ul><li><a href="#요약-abstract">요약 (Abstract)</a></li><li><a href="#도입-배경-rationale">도입 배경 (Rationale)</a></li><li><a href="#제안-proposal">제안 (Proposal)</a><ul><li><a href="#pyerr_-함수들을-수정하여-예외-연결">PyErr_*() 함수들을 수정하여 예외 연결</a></li><li><a href="#예외를-연결하지-않는-함수-수정">예외를 연결하지 않는 함수 수정</a></li><li><a href="#예외를-연결하는-함수-수정">예외를 연결하는 함수 수정</a></li></ul></li><li><a href="#하위-호환성-backward-compatibility">하위 호환성 (Backward compatibility)</a></li><li><a href="#대안-alternatives">대안 (Alternatives)</a><ul><li><a href="#변경-없음-no-change">변경 없음 (No change)</a></li><li><a href="#새로운-예외-연결-헬퍼-함수-new-helpers-to-chain-exceptions">새로운 예외 연결 헬퍼 함수 (New helpers to chain exceptions)</a></li></ul></li><li><a href="#부록-appendix">부록 (Appendix)</a></li><li><a href="#거부-rejection">거부 (Rejection)</a></li><li><a href="#저작권-copyright">저작권 (Copyright)</a></li></ul></li></ul>

            </nav>
          </aside>
        
        <blockquote>
  <p><strong>원문 링크:</strong> <a href="https://peps.python.org/pep-0490/">PEP 490 - Chain exceptions at C level</a></p>

  <table>
    <tbody>
      <tr>
        <td><strong>상태:</strong> Rejected</td>
        <td><strong>유형:</strong> Standards Track</td>
        <td><strong>작성일:</strong> 25-Mar-2015</td>
      </tr>
    </tbody>
  </table>
</blockquote>

<h1 id="pep-490--c-수준에서-예외-연결-chain-exceptions-at-c-level">PEP 490 – C 수준에서 예외 연결 (Chain exceptions at C level)</h1>

<ul>
  <li><strong>작성자:</strong> Victor Stinner <vstinner at="" python.org=""></vstinner></li>
  <li><strong>상태:</strong> Rejected (거부됨)</li>
  <li><strong>유형:</strong> Standards Track</li>
  <li><strong>생성일:</strong> 2015년 3월 25일</li>
  <li><strong>Python 버전:</strong> 3.6</li>
</ul>

<h2 id="요약-abstract">요약 (Abstract)</h2>
<p>이 PEP는 Python 수준에서 이미 예외가 연결(chained)되는 것처럼, C 수준에서도 예외를 연결하는 것을 제안합니다.</p>

<h2 id="도입-배경-rationale">도입 배경 (Rationale)</h2>
<p>Python 3은 기본적으로 예외를 연결(chain)하는 새로운 기능을 도입했습니다 (PEP 3134).</p>

<p>예를 들어, 다음 Python 코드를 보면:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">try</span><span class="p">:</span>
    <span class="k">raise</span> <span class="nc">TypeError</span><span class="p">(</span><span class="sh">"</span><span class="s">err1</span><span class="sh">"</span><span class="p">)</span>
<span class="k">except</span> <span class="nb">TypeError</span><span class="p">:</span>
    <span class="k">raise</span> <span class="nc">ValueError</span><span class="p">(</span><span class="sh">"</span><span class="s">err2</span><span class="sh">"</span><span class="p">)</span>
</code></pre></div></div>
<p>출력은 다음과 같습니다.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Traceback (most recent call last):
  File "test.py", line 2, in &lt;module&gt;
    raise TypeError("err1")
TypeError: err1
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "test.py", line 4, in &lt;module&gt;
    raise ValueError("err2")
ValueError: err2
</code></pre></div></div>
<p>Python 코드에서는 예외가 기본적으로 연결되지만, C로 작성된 확장 모듈에서는 그렇지 않습니다.</p>

<p>새로운 비공개 함수 <code class="language-plaintext highlighter-rouge">_PyErr_ChainExceptions()</code>는 Python 3.4.3 및 3.5에서 예외를 연결하기 위해 도입되었습니다. 현재 이 함수는 예외를 연결하기 위해 명시적으로 호출해야 하며, 그 사용법이 간단하지 않습니다.</p>

<p><code class="language-plaintext highlighter-rouge">zipimport</code> 모듈에서 이전 <code class="language-plaintext highlighter-rouge">OSError</code>를 새로운 <code class="language-plaintext highlighter-rouge">ZipImportError</code> 예외에 연결하기 위해 <code class="language-plaintext highlighter-rouge">_PyErr_ChainExceptions()</code>를 사용하는 예시는 다음과 같습니다.</p>
<div class="language-c highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">PyObject</span> <span class="o">*</span><span class="n">exc</span><span class="p">,</span> <span class="o">*</span><span class="n">val</span><span class="p">,</span> <span class="o">*</span><span class="n">tb</span><span class="p">;</span>
<span class="n">PyErr_Fetch</span><span class="p">(</span><span class="o">&amp;</span><span class="n">exc</span><span class="p">,</span> <span class="o">&amp;</span><span class="n">val</span><span class="p">,</span> <span class="o">&amp;</span><span class="n">tb</span><span class="p">);</span>
<span class="n">PyErr_Format</span><span class="p">(</span><span class="n">ZipImportError</span><span class="p">,</span> <span class="s">"can't open Zip file: %R"</span><span class="p">,</span> <span class="n">archive</span><span class="p">);</span>
<span class="n">_PyErr_ChainExceptions</span><span class="p">(</span><span class="n">exc</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">tb</span><span class="p">);</span>
</code></pre></div></div>
<p>이 PEP는 일관성을 유지하고 디버깅에 도움이 되는 더 많은 실패 정보를 제공하기 위해 C 수준에서도 예외를 자동으로 연결할 것을 제안합니다. 위의 예시는 단순히 다음과 같이 될 수 있습니다.</p>
<div class="language-c highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">PyErr_Format</span><span class="p">(</span><span class="n">ZipImportError</span><span class="p">,</span> <span class="s">"can't open Zip file: %R"</span><span class="p">,</span> <span class="n">archive</span><span class="p">);</span>
</code></pre></div></div>

<h2 id="제안-proposal">제안 (Proposal)</h2>

<h3 id="pyerr_-함수들을-수정하여-예외-연결"><code class="language-plaintext highlighter-rouge">PyErr_*()</code> 함수들을 수정하여 예외 연결</h3>
<p>Python C API에서 예외를 발생시키는 C 함수들(<code class="language-plaintext highlighter-rouge">PyErr_SetString()</code>, <code class="language-plaintext highlighter-rouge">PyErr_Format()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetNone()</code> 등)을 수정하여 예외를 자동으로 연결하도록 제안합니다.</p>

<h3 id="예외를-연결하지-않는-함수-수정">예외를 연결하지 않는 함수 수정</h3>
<p>새로운 예외가 이전 예외의 정보를 포함하거나 더 많은 정보를 포함하는 경우, 특히 두 예외의 유형이 같은 경우에는 이전 예외를 유지하는 것이 항상 유용하지 않을 수 있습니다.</p>

<p><code class="language-plaintext highlighter-rouge">int(str)</code>을 사용한 불필요한 예외 연결의 예시는 다음과 같습니다.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>TypeError: a bytes-like object is required, not 'type'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
TypeError: int() argument must be a string, a bytes-like object or a number, not 'type'
</code></pre></div></div>
<p>새로운 <code class="language-plaintext highlighter-rouge">TypeError</code> 예외가 이전 예외보다 더 많은 정보를 포함하고 있으므로, 이전 예외는 숨겨져야 합니다.</p>

<p>새로운 예외와 현재 예외를 연결하지 않으려면, 새로운 예외를 발생시키기 전에 <code class="language-plaintext highlighter-rouge">PyErr_Clear()</code> 함수를 호출하여 현재 예외를 지울 수 있습니다.</p>

<h3 id="예외를-연결하는-함수-수정">예외를 연결하는 함수 수정</h3>
<p>일부 함수는 현재 예외를 저장한 다음 복원합니다. 새로운 예외가 발생하면, 현재 해당 예외는 함수에 따라 <code class="language-plaintext highlighter-rouge">sys.stderr</code>에 표시되거나 무시됩니다. 이러한 함수 중 일부는 대신 예외를 연결하도록 수정되어야 합니다.</p>

<p>새로운 예외를 무시하는 함수의 예시:</p>
<ul>
  <li><code class="language-plaintext highlighter-rouge">ptrace_enter_call()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">subprocess_fork_exec()</code>: <code class="language-plaintext highlighter-rouge">enable_gc()</code>에 의해 발생하는 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">_thread</code> 모듈의 <code class="language-plaintext highlighter-rouge">t_bootstrap()</code>: 부트스트랩 함수를 <code class="language-plaintext highlighter-rouge">sys.stderr</code>에 표시하려다 발생하는 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">PyDict_GetItem()</code>, <code class="language-plaintext highlighter-rouge">_PyDict_GetItem_KnownHash()</code>: 딕셔너리에서 키를 찾을 때 발생하는 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">_PyErr_TrySetFromCause()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">PyFrame_LocalsToFast()</code>: <code class="language-plaintext highlighter-rouge">dict_to_map()</code>에 의해 발생하는 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">_PyObject_Dump()</code>: 예외 무시 (디버깅 용도로 Python 상태를 수정해서는 안 됨)</li>
  <li><code class="language-plaintext highlighter-rouge">Py_ReprLeave()</code>: 예외 무시 (보고할 방법이 없기 때문)</li>
  <li><code class="language-plaintext highlighter-rouge">type_dealloc()</code>: <code class="language-plaintext highlighter-rouge">remove_all_subclasses()</code>에 의해 발생하는 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">PyObject_ClearWeakRefs()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">call_exc_trace()</code>, <code class="language-plaintext highlighter-rouge">call_trace_protected()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">remove_importlib_frames()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">do_mktuple()</code> (예: <code class="language-plaintext highlighter-rouge">Py_BuildValue()</code>에 사용되는 헬퍼): 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">flush_io()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">sys_write()</code>, <code class="language-plaintext highlighter-rouge">sys_format()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">_PyTraceback_Add()</code>: 예외 무시</li>
  <li><code class="language-plaintext highlighter-rouge">PyTraceBack_Print()</code>: 예외 무시</li>
</ul>

<p>새로운 예외를 <code class="language-plaintext highlighter-rouge">sys.stderr</code>에 표시하는 함수의 예시:</p>
<ul>
  <li><code class="language-plaintext highlighter-rouge">atexit_callfuncs()</code>: <code class="language-plaintext highlighter-rouge">PyErr_Display()</code>로 예외를 표시하고 최신 예외를 반환 (여러 콜백을 호출하고 최신 예외만 반환)</li>
  <li><code class="language-plaintext highlighter-rouge">sock_dealloc()</code>: <code class="language-plaintext highlighter-rouge">ResourceWarning</code> 예외를 <code class="language-plaintext highlighter-rouge">PyErr_WriteUnraisable()</code>로 로깅</li>
  <li><code class="language-plaintext highlighter-rouge">slot_tp_del()</code>: <code class="language-plaintext highlighter-rouge">PyErr_WriteUnraisable()</code>로 예외 표시</li>
  <li><code class="language-plaintext highlighter-rouge">_PyGen_Finalize()</code>: <code class="language-plaintext highlighter-rouge">gen_close()</code> 예외를 <code class="language-plaintext highlighter-rouge">PyErr_WriteUnraisable()</code>로 표시</li>
  <li><code class="language-plaintext highlighter-rouge">slot_tp_finalize()</code>: <code class="language-plaintext highlighter-rouge">__del__()</code> 메서드에 의해 발생하는 예외를 <code class="language-plaintext highlighter-rouge">PyErr_WriteUnraisable()</code>로 표시</li>
  <li><code class="language-plaintext highlighter-rouge">PyErr_GivenExceptionMatches()</code>: <code class="language-plaintext highlighter-rouge">PyType_IsSubtype()</code>에 의해 발생하는 예외를 <code class="language-plaintext highlighter-rouge">PyErr_WriteUnraisable()</code>로 표시</li>
</ul>

<h2 id="하위-호환성-backward-compatibility">하위 호환성 (Backward compatibility)</h2>
<p>예외를 연결하는 것의 부작용 중 하나는 예외가 트레이스백(traceback) 객체를 저장하고, 트레이스백 객체는 프레임(frame) 객체를 저장하며, 프레임 객체는 지역 변수를 저장한다는 것입니다. 지역 변수는 예외에 의해 살아남게 됩니다. 일반적인 문제는 지역 변수와 예외 사이에 순환 참조(reference cycle)가 발생하는 것입니다. 즉, 예외가 지역 변수에 저장되고, 프레임이 예외에 간접적으로 저장되는 경우입니다. 이 순환 참조는 예외를 저장하는 애플리케이션에만 영향을 미칩니다.</p>

<p>이러한 순환 참조는 Python 3.5에 도입된 새로운 <code class="language-plaintext highlighter-rouge">traceback.TracebackException</code> 객체로 해결될 수 있습니다. 이 객체는 지역 변수를 저장하지 않고도 전체 텍스트 트레이스백을 포맷하는 데 필요한 정보를 저장합니다.</p>

<p><code class="language-plaintext highlighter-rouge">asyncio</code> 모듈은 순환 참조 문제의 영향을 받습니다. 이 모듈은 Python 3.3용 버전을 출시하기 위해 Python 표준 라이브러리 외부에서도 유지 관리됩니다. <code class="language-plaintext highlighter-rouge">traceback.TracebackException</code>은 순환 참조 문제를 해결하기 위해 비공개 <code class="language-plaintext highlighter-rouge">asyncio</code> 모듈로 백포트될 수 있습니다.</p>

<h2 id="대안-alternatives">대안 (Alternatives)</h2>

<h3 id="변경-없음-no-change">변경 없음 (No change)</h3>
<p>새로운 비공개 함수 <code class="language-plaintext highlighter-rouge">_PyErr_ChainExceptions()</code>만으로도 예외를 수동으로 연결하기에 충분합니다. 예외는 의미 있는 곳에서만 명시적으로 연결될 것입니다.</p>

<h3 id="새로운-예외-연결-헬퍼-함수-new-helpers-to-chain-exceptions">새로운 예외 연결 헬퍼 함수 (New helpers to chain exceptions)</h3>
<p><code class="language-plaintext highlighter-rouge">PyErr_SetString()</code>과 같은 함수는 예외를 자동으로 연결하지 않습니다. <code class="language-plaintext highlighter-rouge">_PyErr_ChainExceptions()</code>의 사용을 더 쉽게 만들기 위해 다음과 같은 새로운 비공개 함수들이 추가될 수 있습니다.</p>
<ul>
  <li><code class="language-plaintext highlighter-rouge">_PyErr_SetStringChain(exc_type, message)</code></li>
  <li><code class="language-plaintext highlighter-rouge">_PyErr_FormatChain(exc_type, format, ...)</code></li>
  <li><code class="language-plaintext highlighter-rouge">_PyErr_SetNoneChain(exc_type)</code></li>
  <li><code class="language-plaintext highlighter-rouge">_PyErr_SetObjectChain(exc_type, exc_value)</code></li>
</ul>

<p><code class="language-plaintext highlighter-rouge">_PyErr_SetKeyError(key)</code> 또는 <code class="language-plaintext highlighter-rouge">PyErr_SetImportError(message, name, path)</code>와 같은 특정 예외를 발생시키는 헬퍼 함수들은 예외를 연결하지 않습니다. 이 헬퍼 함수들과 예외를 연결하려면 일반적인 <code class="language-plaintext highlighter-rouge">_PyErr_ChainExceptions(exc_type, exc_value, exc_tb)</code> 함수를 사용해야 합니다.</p>

<h2 id="부록-appendix">부록 (Appendix)</h2>

<ul>
  <li><strong>PEPs</strong>:
    <ul>
      <li>PEP 3134 – Exception Chaining and Embedded Tracebacks (Python 3.0): 예외에 대한 새로운 <code class="language-plaintext highlighter-rouge">__context__</code> 및 <code class="language-plaintext highlighter-rouge">__cause__</code> 속성.</li>
      <li>PEP 415 – Implement context suppression with exception attributes (Python 3.3): <code class="language-plaintext highlighter-rouge">raise exc from None</code>.</li>
      <li>PEP 409 – Suppressing exception context (PEP 415로 대체됨).</li>
    </ul>
  </li>
  <li><strong>Python C API</strong>: <code class="language-plaintext highlighter-rouge">Include/pyerror.h</code> 헤더 파일은 예외 관련 함수들을 선언합니다.
    <ul>
      <li>예외 발생 함수: <code class="language-plaintext highlighter-rouge">PyErr_SetNone()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetObject()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetString()</code>, <code class="language-plaintext highlighter-rouge">PyErr_Format()</code>.</li>
      <li>특정 예외 발생 헬퍼: <code class="language-plaintext highlighter-rouge">PyErr_BadArgument()</code>, <code class="language-plaintext highlighter-rouge">PyErr_BadInternalCall()</code>, <code class="language-plaintext highlighter-rouge">PyErr_NoMemory()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetFromErrno()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetFromWindowsErr()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetImportError()</code>, <code class="language-plaintext highlighter-rouge">_PyErr_SetKeyError()</code>, <code class="language-plaintext highlighter-rouge">_PyErr_TrySetFromCause()</code>.</li>
      <li>현재 예외 관리 함수: <code class="language-plaintext highlighter-rouge">PyErr_Clear()</code>, <code class="language-plaintext highlighter-rouge">PyErr_Fetch()</code>, <code class="language-plaintext highlighter-rouge">PyErr_Restore()</code>, <code class="language-plaintext highlighter-rouge">PyErr_GetExcInfo()</code>, <code class="language-plaintext highlighter-rouge">PyErr_SetExcInfo()</code>.</li>
      <li>예외 처리 기타 함수: <code class="language-plaintext highlighter-rouge">PyErr_ExceptionMatches()</code>, <code class="language-plaintext highlighter-rouge">PyErr_GivenExceptionMatches()</code>, <code class="language-plaintext highlighter-rouge">PyErr_NormalizeException()</code>, <code class="language-plaintext highlighter-rouge">_PyErr_ChainExceptions()</code>.</li>
    </ul>
  </li>
  <li><strong>Python Issues</strong>: 예외 연결 및 예외 손실 방지와 관련된 여러 이슈가 언급되어 있습니다.</li>
</ul>

<h2 id="거부-rejection">거부 (Rejection)</h2>
<p>이 PEP는 2017년 9월 12일 Victor Stinner에 의해 거부되었습니다. <code class="language-plaintext highlighter-rouge">python-dev</code> 논의에서 C 예외를 기본적으로 연결하지 않고, 대신 의미 있는 곳에서만 명시적으로 연결하기로 결정되었습니다.</p>

<h2 id="저작권-copyright">저작권 (Copyright)</h2>
<p>이 문서는 퍼블릭 도메인에 공개되었습니다.</p>

<blockquote>
  <p>⚠️ <strong>알림:</strong> 이 문서는 AI를 활용하여 번역되었으며, 기술적 정확성을 보장하지 않습니다. 정확한 내용은 반드시 원문을 확인하시기 바랍니다.</p>
</blockquote>

        
      </section>

      <footer class="page__meta">
        
        
  


  

  <p class="page__taxonomy">
    <strong><i class="fas fa-fw fa-tags" aria-hidden="true"></i> Tags: </strong>
    <span itemprop="keywords">
    
      <a href="/tags/#pep" class="page__taxonomy-item" rel="tag">PEP</a><span class="sep">, </span>
    
      <a href="/tags/#python" class="page__taxonomy-item" rel="tag">Python</a><span class="sep">, </span>
    
      <a href="/tags/#translation" class="page__taxonomy-item" rel="tag">Translation</a>
    
    </span>
  </p>




  


  

  <p class="page__taxonomy">
    <strong><i class="fas fa-fw fa-folder-open" aria-hidden="true"></i> Categories: </strong>
    <span itemprop="keywords">
    
      <a href="/categories/#pep" class="page__taxonomy-item" rel="tag">PEP</a><span class="sep">, </span>
    
      <a href="/categories/#python" class="page__taxonomy-item" rel="tag">Python</a>
    
    </span>
  </p>


        
  <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2025-09-26">September 26, 2025</time></p>


      </footer>

      

      
  <nav class="pagination">
    
      <a href="/python/pep/489/" class="pagination--pager" title="[Final] PEP 489 - Multi-phase extension module initialization
">Previous</a>
    
    
      <a href="/python/pep/491/" class="pagination--pager" title="[Deferred] PEP 491 - The Wheel Binary Package Format 1.9
">Next</a>
    
  </nav>


    </div>

    
  </article>

</div>

    </div>

    
      <div class="search-content">
        <div class="search-content__inner-wrap"><form class="search-content__form" onkeydown="return event.key != 'Enter';">
    <label class="sr-only" for="search">
      Enter your search term...
    </label>
    <input type="search" id="search" class="search-input" tabindex="-1" placeholder="Enter your search term..." />
  </form>
  <div id="results" class="results"></div></div>

      </div>
    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    

    
      
        
      
        
      
        
      
        
      
        
      
        
      
    

    
      <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
    
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2025 secrett2633. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>




<script src="/assets/js/lunr/lunr.min.js"></script>
<script src="/assets/js/lunr/lunr-store.js"></script>
<script src="/assets/js/lunr/lunr-en.js"></script>




    <script>
  'use strict';

  (function() {
    var commentContainer = document.querySelector('#utterances-comments');

    if (!commentContainer) {
      return;
    }

    var script = document.createElement('script');
    script.setAttribute('src', 'https://utteranc.es/client.js');
    script.setAttribute('repo', 'secrett2633/secrett2633.github.io');
    script.setAttribute('issue-term', 'pathname');
    script.setAttribute('theme', 'github-light');
    script.setAttribute('crossorigin', 'anonymous');

    commentContainer.appendChild(script);
  })();
</script>

  





  </body>
</html>
